test/functional/test-shell.md

```diff
@@ -0,0 +1,188 @@
+Test Shell for Interactive Environments
+=========================================
+
+This document describes how to use the `TestShell` submodule in the functional
+test suite.
+
+The `TestShell` submodule extends the `BitcoinTestFramework` functionality to
+external interactive environments for prototyping and educational purposes. Just
+like `BitcoinTestFramework`, the `TestShell` allows the user to:
+
+* Manage regtest bitcoind subprocesses.
+* Access RPC interfaces of the underlying bitcoind instances.
+* Log events to the functional test logging utility.
+
+The `TestShell` can be useful in interactive environments where it is necessary
+to extend the object lifetime of the underlying `BitcoinTestFramework` between
+user inputs. Such environments include the Python3 command line interpreter or
+[Jupyter](https://jupyter.org/) notebooks running a Python3 kernel.
+
+## 1. Requirements
+
+* Python3
+* `bitcoind` built in the same repository as the `TestShell`.
+
+## 2. Importing `TestShell` from the Bitcoin Core repository
+
+We can import the `TestShell` by adding the path of the Bitcoin Core
+`test_framework` module to the beginning of the PATH variable, and then
+importing the `TestShell` class from the `test_shell` sub-package.
+
+```
+>>> import sys
+>>> sys.path.insert(0, "/path/to/bitcoin/test/functional")
+>>> from test_framework.test_shell import `TestShell`
```

Comment:
```
                                          ^
SyntaxError: invalid syntax

Reply:
Sorry. I'll follow-up with a fix.

Reply:
While using TestShell from test/functional allows skipping the `sys.path` step, for the common use case of calling TestShell from the repo root it may be helpful to provide a concrete example in the doc:
```
$ python3
Python 3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path.insert(0, "test/functional/")
>>> from test_framework.test_shell import TestShell
>>> test = TestShell()
>>> test.setup(num_nodes=2, setup_clean_chain=True)
2019-11-04T21:13:37.384000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_hrgo_cv4
>>> test.nodes[0].getblockchaininfo()["blocks"]
0
>>> test.shutdown()
2019-11-04T21:13:52.920000Z TestFramework (INFO): Stopping nodes
2019-11-04T21:13:53.333000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_hrgo_cv4 on exit
2019-11-04T21:13:53.334000Z TestFramework (INFO): Tests successful
>>> exit()
```

The reason is that the `sys.path` call from root works with "test/functional/" or "test/functional", but not "/test/functional" (at least for me on debian), which people might try from seeing "/path/to/bitcoin/test/functional" in the current example:
```
>>> import sys
>>> sys.path.insert(0, "/test/functional")
>>> from test_framework.test_shell import TestShell
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'test_framework'
```
Just a thought, YMMV.